devito/ir/clusters/algorithms.py

```diff
@@ -370,34 +396,113 @@ def fuse(clusters):
     return processed
 
 
-def guard(clusters):
+def scalarize(clusters, template):
```

Comment:
note for reviewers: this is pretty much what was in dse/promotion.py, just moved here as it's a "cross-cluster" optimization (while the DSE works on intra-cluster optimizations)